databuilder/query_language.py

```diff
@@ -32,12 +32,25 @@ def compile(self) -> dict[str, Node]:  # noqa A003
 
 
 class PatientTable:
-    pass
+    def __init__(self):
+        self.qm_node = SelectPatientTable(name=self.__name__)
 
 
 class Series:
-    def __init__(self, qm_node):
-        self.qm_node = qm_node
+    @classmethod
+    def factory(cls, qm_node):
+        # We use a factory method in preference to __init__ because we don't want
+        # self.qm_node to be None.
+        instance = cls()
+        instance.qm_node = qm_node
+        return instance
```

Comment:
How does this stop `qm_mode` being None?  Couldn't one pass in None here just like in an `__init__`?

Reply:
I worded my comment poorly. I wanted to avoid:

```python
class Series:
    def __init__(self, qm_node=None):
        self.qm_node = qm_node
```

Which is potentially confusing, because `Series.qm_node` will never be `None` when `Series` is used as a descriptor.

We could, indeed, pass `None` here -- but we could error, too.

Reply:
> I just haven't yet written the error code

Ah, that's where my confusion came from!

FWIW I don't see what `.factory()` with a check for `qm is None` gives you over an `__init__()` with the same check 🙂

Reply:
Yup, agreed. I've replaced `factory` with `__init__` and it looks like it should be easier to reason about, at least to me 🙂 For one thing, `my_series.qm_node` will now always exist, rather than existing in some cases but not in others.